sympy/codegen/ast.py

```diff
@@ -0,0 +1,732 @@
+    directly, like AddAugmentedAssignment(x, y).
+
+    Parameters
+    ----------
+    lhs : Expr
+        Sympy object representing the lhs of the expression. These should be
+        singular objects, such as one would use in writing code. Notable types
+        include Symbol, MatrixSymbol, MatrixElement, and Indexed. Types that
+        subclass these types are also supported.
+
+    op : str
+        Operator (+, -, /, *, %).
+
+    rhs : Expr
+        Sympy object representing the rhs of the expression. This can be any
+        type, provided its shape corresponds to that of the lhs. For example,
+        a Matrix type can be assigned to MatrixSymbol, but not to Symbol, as
+        the dimensions will not align.
+
+    Examples
+    --------
+
+    >>> from sympy import symbols
+    >>> from sympy.codegen.ast import aug_assign
+    >>> x, y = symbols('x, y')
+    >>> aug_assign(x, '+', y)
+    AddAugmentedAssignment(x, y)
+    """
+    if op + '=' not in Relational.ValidRelationOperator:
+        raise ValueError("Unrecognized operator %s" % op)
+    return Relational.ValidRelationOperator[op + '='](lhs, rhs)
+
+
+class CodeBlock(Basic):
+    """
+    Represents a block of code
+
+    For now only assignments are supported. This restriction will be lifted in
+    the future.
+
+    Useful methods on this object are
+
+    ``left_hand_sides``: Tuple of left-hand sides, in order.
+    ``left_hand_sides``: Tuple of right-hand sides, in order.
+    ``topological_sort``: Class method. Return a CodeBlock with assignments
+                          sorted so that variables are assigned before they
+                          are used.
+    ``cse``: Return a new CodeBlock with common subexpressions eliminated and
+             pulled out as assignments.
+
+    Example
+    =======
+
+    >>> from sympy import symbols, ccode
+    >>> from sympy.codegen.ast import CodeBlock, Assignment
+    >>> x, y = symbols('x y')
+
+    >>> c = CodeBlock(Assignment(x, 1), Assignment(y, x + 1))
+    >>> print(ccode(c))
+    x = 1;
+    y = x + 1;
+    """
+    def __new__(cls, *assignments):
+        if not all(isinstance(i, Assignment) for i in assignments):
+            # Will support more things later
+            raise TypeError("CodeBlock inputs must be Assignments")
+
+        left_hand_sides = []
+        right_hand_sides = []
+        for i in assignments:
+            lhs, rhs = i.args
+            if lhs in left_hand_sides:
+                raise NotImplementedError("Duplicate assignments to the same "
+                "variable are not yet supported (%s)" % lhs)
+            left_hand_sides.append(lhs)
+            right_hand_sides.append(rhs)
+
+        obj = Basic.__new__(cls, *assignments)
+
+        obj.left_hand_sides = Tuple(*left_hand_sides)
+        obj.right_hand_sides = Tuple(*right_hand_sides)
+
+        return obj
+
+    @classmethod
+    def topological_sort(cls, assignments):
```

Comment:
If you expect this not to work with augmented assignments perhaps we should raise a `NotImplementedError` or at least mention in the docstring that the behaviour is undefined?